public/externalLibs/sound/sounds.js

```diff
@@ -1,490 +1,501 @@
 // Constants
-var FS = 32000 // Standard sampling rate for all problems
+var FS = 44100; // Standard sampling rate for all problems
 
 // ---------------------------------------------
 // Fast reimplementations of the list library
 // ---------------------------------------------
+function vector_to_list(arr) {
+    var xs = [];
+
+    for (var i=arr.length-1; i>=0; i--) {
+        xs = pair(arr[i], xs);
+    }
+
+    return xs;
+}
+
+function list_to_vector(xs) {
+    var vector = [];
+
+    while(!is_null(xs)) {
+        vector.push(head(xs));
+        xs = tail(xs);
+    }
+
+    return vector;
+}
+
+function length(xs) {
+    var len = 0;
+    while(!is_null(xs)) {
+        len++;
+        xs = tail(xs);
+    }
+    return len;
+}
 
 function append(xs, ys) {
-  var v1 = list_to_vector(xs)
-  var v2 = list_to_vector(ys)
-  var vector = v1.concat(v2)
-  return vector_to_list(vector)
+    var v1 = list_to_vector(xs);
+    var v2 = list_to_vector(ys);
+    var vector = v1.concat(v2);
+    return vector_to_list(vector);
 }
 
 function map(f, xs) {
-  var vector = list_to_vector(xs)
-  for (var i = 0; i < vector.length; i++) {
-    vector[i] = f(vector[i])
-  }
-  return vector_to_list(vector)
+    var vector = list_to_vector(xs);
+    for (var i=0; i<vector.length; i++) {
+        vector[i] = f(vector[i]);
+    }
+    return vector_to_list(vector);
 }
 
 // ---------------------------------------------
 // Low-level sound support
 // ---------------------------------------------
 
-// samples a continuous wave to a discrete waves at sampling rate for duration
+// Samples a continuous wave to a discrete waves at sampling rate for duration
 // in seconds
 function discretize(wave, duration) {
-  var vector = []
+    var vector = [];
 
-  for (var i = 0; i < duration * FS; i++) {
-    vector.push(wave(i / FS))
-  }
+    for (var i = 0; i < duration * FS; i++) {
+        vector.push(wave( i / FS));
+    }
 
-  return vector
+    return vector;
+}
+
+// Discretizes a sound to a sound starting from elapsed_duration, for sample_length seconds
+function discretize_from(wave, duration, elapsed_duration, sample_length, data) {
```

Comment:
Cant you make this into one for loop? (You jus need to change the terminating condition)

e.g.
```javascript
var completion = (elapsed_duration + sample_length > duration) ? duration * FS : (elapsed_duration + sample_length) * FS;
for (var i = elapsed_duration * FS; i < completion; i++) {
    data[i - elapsed_duration * FS] = wave(i / FS); 
}
return data;
```